man/ws_allocate.1
.TH ws_allocate 1 "March 2013" "USER COMMANDS"

.SH NAME
ws_allocate \- create a workspace directory and return its path

.SH SYNOPSIS
.B ws_allocate
[\-h] [\-x] [\-g] [\-G GROUPNAME] [\-F FILESYSTEM] [\-r DAYS] [\-m MAILADDRESS] [\-c COMMENT] NAME DURATION

.SH DESCRIPTION
Create a 
.B workspace 
for the calling user with name NAME which will exist for DURATION days.
The command will always return the full name of the created directory on stdout, and will print
some information to stderr.
A 
.B workspace 
has a limited lifetime, which can be restricted by the administrator.
The DURATION might be cut to that maximum, a warning will be printed in that case.

A 
.B workspace 
can be extended if allowed by the administrator, who might limit the number
of extensions.

A reminder email can be sent to address MAILADDRESS (or address from ~/.ws_user.conf)
DAYS before the 
.B
workspace 
expires. This will done probably during a nightly cleanup run, do not expect it to come
exactly in time but rather the night after the warning period is met.

The idea of the 
.B workspace 
is to allow the administrator to keep control of filespace in working filesystems,
and to avoid the situations that data of past and finished projects/jobs is idling for
ever on disks. A user can create and dispose arbitrary number of workspaces, they are cheap,
and can be created e.g. for each job.

For the administrator, 
.B workspaces 
offer further advantages, like beeing able to migrate users to another filesystem over time,
and to do loadbalancing over directories and filesystems.

From the system side, 
.B workspaces 
will be deleted after beeing expired. A friendly user can release workspaces before the expiration
date using
.B ws_release
which will mark the 
.B workspace 
for future deletion and make the data unavailable for the user.

Deletion is probably a task done once a night and not immediately.

If a
.B workspace
was forgotten and expired, it might make sense to contact the administrator,
as he probaly can restore a workspace for some time, if disk space is not low.
Check if 
.B ws_restore
is available to restore workspaces that are still in the hidden area.

.PP

.SH OPTIONS
.TP
\-h 
display usage help text
.TP
\-x
extend an existing workspace until today + DURATION days.
The number of extensions can be limited. Can be combined with -m to add a mail address
and with -r to add reminder to existing workspaces. If used with 0 as duration, it does
not consume an extension.
.TP
\-F
select the filesystem to create the workspace in, there is always a default for this.
See 
.B ws_list
\-l 
for a list of available filesystems.
.TP
\-m
mailaddress to send reminder mail to, can be combined with -x to add mailaddress to existing workspace (combine with duration 0).
.TP
\-c
comment to attach to workspace, can be combined with -x to change comment of an existing workspace (combine with duration 0).
.TP
\-r DAYS
reminder email will be sent DAYS days before the
.B
workspace 
will expire, can be combined with -x to add reminder to existing workspace. This could be sent in the night after the exact point in time, during a cleanup.
See 
.B ws_send_ical 
for a way to send calendar entries.
.TP
\-g 
a group-workspace will be created, that is a workspace that can be seen by other members of current group with 
.B ws_list -g
and the workspace itself is created group readable.
.TP
\-G GROUPNAME
use the GROUPNAME as group for the workspace, make the workspace group writable and set the group sticky bit.

.SH EXAMPLES
.TP
create a workspace for 10 days:
.B ws_allocate
myworkspace 10

.SH FILES
.B
~/.ws_user.conf 
user config file to store some default values. file format is yaml, possible keys are mail, duration, reminder and groupname.
Those values can be overwritten with options on commandline.

.TP
Example:
mail: user@domain
.br
reminder: 1
.br
groupname: boygroup
.br
duration: 10


.SH AUTHOR
Written by Holger Berger

.SH SEE ALSO
ws_release, ws_list, ws_find, ws_restore, ws_send_ical


debian/changelog
hpc-workspace (1.3.6-1debian) bullseye; urgency=medium

  * Initial release

 -- Markus Kötter <koetter@cispa.de>  Thu, 09 Dec 2021 07:40:07 +0100


src/ws_allocate.cpp
/*
 *  workspace++
 *
 *  ws_allocate
 *
 *  c++ version of workspace utility
 *  a workspace is a temporary directory created in behalf of a user with a limited lifetime.
 *  This version is not DB and configuration compatible with the older version, the DB and 
 *  configuration was changed to YAML files.
 * 
 *  differences to old workspace version
 *    - usage of YAML file format
 *    - using setuid or capabilities (needs support by filesystem!)
 *    - supports configuration of reminder emails
 *
 *  (c) Holger Berger 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022
 * 
 *  workspace++ is based on workspace by Holger Berger, Thomas Beisel and Martin Hecht
 *
 *  workspace++ is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  workspace++ is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with workspace++.  If not, see <http://www.gnu.org/licenses/>.
 *
 */


#include <iostream>
#include <fstream>
#include <string>
#ifdef USE_BOOST_REGEXP
	#include <boost/regex.hpp>
    #define REGEX boost::regex
#else
	#include <regex>
    #define REGEX std::regex
#endif
#include <syslog.h>

// YAML
#include <yaml-cpp/yaml.h>

#include <boost/program_options.hpp>
#include <boost/filesystem.hpp>

#ifndef SETUID
#include <sys/capability.h>
#else
typedef int cap_value_t;
const int CAP_DAC_OVERRIDE = 0;
const int CAP_CHOWN = 1;
#endif

#include "ws.h"

namespace po = boost::program_options;
using namespace std;


/* 
 *  parse the commandline and see if all required arguments are passed, and check the workspace name for 
 *  bad characters
 */
void commandline(po::variables_map &opt, string &name, int &duration, const int durationdefault, string &filesystem, 
                    bool &extension, int &reminder, string &mailaddress, string &user, string &groupname, string &comment,
                    int argc, char**argv, std::stringstream &userconf) {
    // define all options

    po::options_description cmd_options( "\nOptions" );
    cmd_options.add_options()
            ("help,h", "produce help message")
            ("version,V", "show version")
            ("duration,d", po::value<int>(&duration), "duration in days")
            //("duration,d", po::value<int>(&duration)->default_value(durationdefault), "duration in days")
            ("name,n", po::value<string>(&name), "workspace name")
            ("filesystem,F", po::value<string>(&filesystem), "filesystem")
            ("reminder,r", po::value<int>(&reminder), "reminder to be sent n days before expiration")
            ("mailaddress,m", po::value<string>(&mailaddress), "mailaddress to send reminder to")
            ("extension,x", "extend workspace")
            ("username,u", po::value<string>(&user), "username")
            ("group,g", "group workspace")
            ("groupname,G", po::value<string>(&groupname)->default_value(""), "groupname")
            ("comment,c", po::value<string>(&comment), "comment")
    ;

    po::options_description secret_options("Secret");
    secret_options.add_options()
        ("debug", "show debugging information")
        ;

    // define options without names
    po::positional_options_description p;
    p.add("name", 1).add("duration",2);

    po::options_description all_options;
    all_options.add(cmd_options).add(secret_options);

    // parse commandline
    try{
        po::store(po::command_line_parser(argc, argv).options(all_options).positional(p).run(), opt);
        po::notify(opt);
    } catch (...) {
        cout << "Usage: " << argv[0] << ": [options] workspace_name duration" << endl;
        cout << cmd_options << "\n";
        exit(1);
    }

    // see whats up

    if (opt.count("help")) {
        cout << "Usage: " << argv[0] << ": [options] workspace_name duration" << endl;
        cout << cmd_options << "\n";
        exit(1);
    }

    if (opt.count("version")) {
#ifdef IS_GIT_REPOSITORY
        cout << "workspace build from git commit hash " << GIT_COMMIT_HASH
             << " on top of release " << WS_VERSION << endl;
#else
        cout << "workspace version " << WS_VERSION << endl;
#endif
        exit(1);
    }

    // this allows user to extend foreign workspaces
    if(opt.count("username") && !( opt.count("extension") || getuid()==0 ) ) {
        cerr << "Info: Ignoring username option." << endl;
		user="";
    }

    if(opt.count("extension")) {
        extension = true;
    } else {
        extension = false;
    }

    if (opt.count("name"))
    {
        //cout << " name: " << name << "\n";
    } else {
        cout << argv[0] << ": [options] workspace_name duration" << endl;
        cout << cmd_options << "\n";
        exit(1);
    }
	
	// get some values from ~/.ws_user.conf if not on commandline
	//  mail:  mail address for reminder
	//  reminder: days before expiry to send reminder
	//  duration: duration of workspace 
	//  groupname: group for group workspace
	YAML::Node user_home_config;
	std::string firstline;
    getline(userconf, firstline);
    if ((firstline.find(":",0) == string::npos) && (firstline.find("@",0) != string::npos)) {
		cerr << "Info: ignoring ~/.ws_user.conf, seems to be in old, non yaml format and contain no mail: in front of mail address" << endl;
	} else {
        user_home_config = YAML::Load(userconf.str());
		if(opt.count("reminder")==0) {
			try {
				reminder = user_home_config["reminder"].as<int>();
				cerr << "Info: reminder <" << reminder << "> taken from ~/.ws_user.conf" << endl;
			} catch (...) {
				reminder = 0;
			}
		}
		if(groupname=="") {
			try {
				groupname = user_home_config["groupname"].as<std::string>();
				// ugly hack to get changed groupname into validator, we parse some pseudo commandline fake command line option
				auto tg = groupname;
				int largc=3; char* largv[] = {(char *)"",(char*)"-G",(char*)tg.c_str(),NULL};
				po::store(po::command_line_parser(largc, largv).options(all_options).positional(p).run(), opt);
				po::notify(opt);
				cerr << "Info: groupname <" << groupname << "> taken from ~/.ws_user.conf" << endl;
				cerr << "Warning: your workspace will be group writable, as your ~/.ws_user.conf contains a groupname directive!" << endl;
			} catch (...) {
				groupname = "";
			}
		}
		if(opt.count("duration")==0) {
			try {
				duration = user_home_config["duration"].as<int>();
				cerr << "Info: duration <" << duration << "> taken from ~/.ws_user.conf" << endl;
			} catch (...) {
				duration = durationdefault;
			}
		} 
	}

    // reminder check, if we have a reminder number, we need either a mailaddress argument or config file
    // with mailaddress in user home
    if(reminder!=0) {
        if (!opt.count("mailaddress")) {
            // check if file looks like yaml
            if (firstline.find(":",0) != string::npos) {
                try {
                    mailaddress = user_home_config["mail"].as<std::string>();
                } catch (...) {
                    mailaddress = "";
                }
            } else {
				mailaddress = firstline;
			}
            if(mailaddress.length()>0) {
                cerr << "Info: Took email address <" << mailaddress << "> from users config." << endl;
            } else {
                mailaddress = Workspace::getusername();
                cerr << "Info: could not read email from users config ~/.ws_user.conf." << endl;
                cerr << "Info: reminder email will be sent to local user account" << endl;
            }
        }
		if (reminder>=duration) {
                cerr << "Warning: reminder is only sent after workspace expiry!" << endl;
		}
    } else {
        // check if mail address was set with -m but not -r
        if(opt.count("mailaddress") && !opt.count("extension")) {
            cerr << "Error: You can't use the mailaddress (-m) without the reminder (-r) option." << endl;
            exit(1);
        } 
    }


    // validate workspace name against nasty characters    
    //  static const std::regex e("^[a-zA-Z0-9][a-zA-Z0-9_.-]*$");  // #77

	// bugfix: as regexp parser are recursiv, split in two parts, complex match for start, simple search for remainder
	static const REGEX e1("^[[:alnum:]][[:alnum:]_.-]*$");
    if (!regex_match(name.substr(0,2) , e1)) {
            cerr << "Error: Illegal workspace name, use characters and numbers, -,. and _ only!" << endl;
            exit(1);
    }
	static const REGEX e2("[^[:alnum:]_.-]");
    if (regex_search(name, e2)) {
            cerr << "Error: Illegal workspace name, use characters and numbers, -,. and _ only!" << endl;
            exit(1);
    }

}



/*
 *  main logic here
 */

int main(int argc, char **argv) {
    int duration, durationdefault;
    bool extensionflag;
    string name;
    string filesystem;
    string mailaddress("");
    string user_option, groupname;
	string comment;
    int reminder = 0;
    po::variables_map opt;

    // we only support C locale, if the used local is not installed on the system
    // ws_allocate fails
    setenv("LANG","C",1);
    setenv("LC_CTYPE","C",1);
	setenv("LC_ALL","C",1);
    std::setlocale(LC_ALL, "C");
    std::locale::global(std::locale("C"));
	boost::filesystem::path::imbue(std::locale());
    
    // read config (for dbuid, reminder and duration default only)
    YAML::Node config;
    try {
        config = YAML::LoadFile("/etc/ws.conf");
    } catch (const YAML::BadFile& e) {
        cerr << "Error: Could not read config file!" << endl;
        cerr << e.what() << endl;
        exit(-1);
    }

    reminder = config["reminderdefault"].as<int>(0);
	durationdefault = config["durationdefault"].as<int>(1);

    int db_uid = config["dbuid"].as<int>();

    // read user config before dropping privileges to DB user
    //
    // get user first, so user config is read as owner
    // of files, which is needed for root_squash homes
    Workspace::drop_cap(CAP_DAC_OVERRIDE, CAP_CHOWN, getuid());

    std::stringstream user_conf;
    string user_conf_filename = Workspace::getuserhome()+"/.ws_user.conf";
    if (!boost::filesystem::is_symlink(user_conf_filename)) {
        std::ifstream t(user_conf_filename.c_str());
        user_conf << t.rdbuf();
    } else {
        cerr << "Error: ~/.ws_user.conf can not be symlink!" << endl;
        exit(-1);
    }

    Workspace::raise_cap(CAP_DAC_OVERRIDE);

    // lower capabilities to minimum
    Workspace::drop_cap(CAP_DAC_OVERRIDE, CAP_CHOWN, db_uid);

    // check commandline, get flags which are used to create ws object or for workspace allocation
    commandline(opt, name, duration, durationdefault , filesystem, extensionflag, 
				reminder, mailaddress, user_option, groupname, comment, argc, argv, user_conf);

    openlog("ws_allocate", 0, LOG_USER); // SYSLOG

    // get workspace object
    Workspace ws(WS_Allocate, opt, duration, filesystem);
    
    // allocate workspace
    ws.allocate(name, extensionflag, reminder, mailaddress, user_option, groupname, comment);
    
}
